ext/random/random.c

```diff
@@ -0,0 +1,871 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Go Kudo <zeriyoshi@gmail.com>                                |
+   +----------------------------------------------------------------------+
+ */
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "php.h"
+#include "ext/standard/info.h"
+#include "ext/standard/php_random.h"
+#include "zend_exceptions.h"
+#include "zend_interfaces.h"
+
+#include "php_random_module.h"
+#include "random_arginfo.h"
+
+PHPAPI zend_class_entry *php_random_ce_Random_NumberGenerator_RandomNumberGenerator;
+PHPAPI zend_class_entry *php_random_ce_Random_NumberGenerator_XorShift128Plus;
+PHPAPI zend_class_entry *php_random_ce_Random_NumberGenerator_MT19937;
+PHPAPI zend_class_entry *php_random_ce_Random_NumberGenerator_Secure;
+PHPAPI zend_class_entry *php_random_ce_RandomInterface;
+PHPAPI zend_class_entry *php_random_ce_Random;
+
+static zend_object_handlers php_random_rng_xorshift128plus_object_handlers;
+static zend_object_handlers php_random_rng_mt19937_object_handlers;
+static zend_object_handlers php_random_rng_secure_object_handlers;
+static zend_object_handlers php_random_php_random_object_handlers;
+
+static inline uint64_t php_random_rng_next(php_random_rng *rng, bool truncate)
+{
+	uint64_t ret;
+
+	ret = rng->algo->generate(rng->state);
+
+	if (truncate) {
+		ret = (zend_long) (ret >> 1);
+	}
+
+	return ret;
+}
+
+uint64_t php_random_next(php_random *random, bool truncate)
+{
+	zval retval;
+	zend_object *rng;
+	zend_string *rng_fname;
+	zend_function *rng_func;
+
+	if (random->rng) {
+		return php_random_rng_next(random->rng, truncate);
+	}
+
+	rng = Z_OBJ_P(zend_read_property(random->std.ce, &random->std, "rng", sizeof("rng") - 1, 0, NULL));
```

Comment:
Should add a check that the zval from zend_read_property is IS_OBJECT.

I'm not sure if using a property is efficient, though - e.g. adding a `zend_object *obj` to `php_random` instead would make that a simple pointer lookup.

Also, with Closure->bindTo, it's possible for userland code to call unset($this->std), I think?

For the alternative, it's possible to override get_properties, get_gc, etc. handlers. Look at SplDoublyLinkedList as an example in ext/spl/spl_dll.c

Reply:
Thank you. I'll take a look.